contracts/lens/FeeOnTransfer.sol

```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+pragma solidity =0.7.6;
+pragma abicoder v2;
+
+import '@uniswap/v3-periphery/contracts/base/PeripheryImmutableState.sol';
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Callee.sol';
+import '@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol';
+import '../libraries/UniswapV2Library.sol';
+import '../interfaces/ISwapRouter02.sol';
+import '../interfaces/IFeeOnTransfer.sol';
+import '../interfaces/IERC20.sol';
+import '../interfaces/IWETH.sol';
+
+/// @notice Determine if a token takes a fee on transfer by flash borrowing from the token/<base token> pool on V2.
+/// @notice Returns true if we detected a fee is taken on transfer.
+/// @notice A return value of false does not mean the token is definitely not a fee on transfer token. It means
+/// @notice we are *unsure* if the token will take a fee on transfer.
+/// @dev We can not guarantee the result of this lens is correct for a few reasons:
+/// @dev 1/ Some tokens take fees under specific conditions, for example some have an allowlist
+/// @dev    of addresses that do/dont require fees. Therefore the result is not guaranteed to be correct
+/// @dev    in all circumstances.
+/// @dev 2/ It is possible that the token does not have any pools on V2 therefore we are not able to perform
+/// @dev    a flashloan to test the transfer.
+contract FeeOnTransfer is IFeeOnTransfer, IUniswapV2Callee, PeripheryImmutableState {
+    string internal constant FOT_REVERT_STRING = 'FOT';
+    bytes4 internal constant PAIR_TOKEN_0_SELECTOR = bytes4(keccak256('token0()'));
+
+    constructor(address _factory, address _WETH9) PeripheryImmutableState(_factory, _WETH9) {}
+
+    function batchIsFeeOnTransfer(
+        address[] calldata tokens,
+        address[] calldata baseTokens,
+        uint256 amountToBorrow
+    ) public override returns (bool[] memory isFotResults) {
+        isFotResults = new bool[](tokens.length);
+        for (uint256 i = 0; i < tokens.length; i++) {
+            isFotResults[i] = isFeeOnTransfer(tokens[i], baseTokens, amountToBorrow);
+        }
+    }
+
+    function isFeeOnTransfer(
+        address token,
+        address[] calldata baseTokens,
+        uint256 amountToBorrow
+    ) public override returns (bool) {
+        for (uint256 i = 0; i < baseTokens.length; i++) {
+            if (_isFeeOnTransfer(token, baseTokens[i], amountToBorrow)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    function _isFeeOnTransfer(
+        address token,
+        address baseToken,
+        uint256 amountToBorrow
+    ) internal returns (bool) {
+        if (token == baseToken) {
+            return false;
+        }
+
+        address pairAddress = UniswapV2Library.pairFor(this.factory(), token, baseToken);
```

Comment:
ah, so this is the v3 factory (from `PeripheryImmutableState`), you should also inherit from `ImmutableState` to get the v2 factory, for consistency